### PR TITLE
Refactor ApiGenerator

### PR DESCRIPTION
**Pull request by Nyholm**
This is a massive rewrite of the ApiGenerator. I've split the ApiGenerator up to three files: 
- OperationGenerator
- ResultGenerator
- GeneratorHelper

I also created some objects from our Shape and Operation. I based it on @jderusse idea:

```php
$operation = $definition->getOperation();
$shape = $operation->getOutput();
if ($shape instanceOf StructureShape) {
  echo count($shape->getMembers());
} elseif ($shape instanceOf ListShape) {
  echo get_class($shape->getMember());
} else {
  echo $shape->getType();
}
```

Only files in src/CodeGenerator was changed. So we know the result of my refactoring did not effect the output.

### CODE REVIEW

**Comment by Nyholm** [comment]
Future work: One could of course use the `Shape` and `Operation` classes more. I've cheated by letting them implement `\ArrayAccess`

**Comment by Nyholm** [comment]
Im done with the PR now. 

The final push extracted XmlParser form ResultGenerator

File: src/CodeGenerator/Generator/GeneratorHelper.php [hunk 1]
```diff
@@ -12,6 +12,7 @@
  * Small methods that might be useful when generating code.
  *
  * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ * @author Jérémy Derussé <jeremy@derusse.com>
```
Comment (jderusse) [inline src/CodeGenerator/Generator/GeneratorHelper.php]: thanks :hugs:

Comment (Nyholm) [reply src/CodeGenerator/Generator/GeneratorHelper.php]: We should remove all or slap your name all over the place.

**Pull request merged**